swift/ql/lib/codeql/swift/dataflow/internal/DataFlowPrivate.qll

```diff
@@ -785,6 +827,156 @@ private module OutNodes {
 
 import OutNodes
 
+/**
+ * Holds if there is a data flow step from `e1` to `e2` that only steps from
+ * child to parent in the AST.
+ */
+private predicate simpleAstFlowStep(Expr e1, Expr e2) {
+  e2.(IfExpr).getBranch(_) = e1
+  or
+  e2.(AssignExpr).getSource() = e1
+  or
+  e2.(ArrayExpr).getAnElement() = e1
+}
+
+private predicate closureFlowStep(CaptureInput::Expr e1, CaptureInput::Expr e2) {
+  simpleAstFlowStep(e1, e2)
+  or
+  exists(Ssa::WriteDefinition def |
+    def.getARead().getNode().asAstNode() = e2 and
+    def.assigns(any(CfgNode cfg | cfg.getNode().asAstNode() = e1))
+  )
+  or
+  e2.(Pattern).getImmediateMatchingExpr() = e1
+}
+
+private module CaptureInput implements VariableCapture::InputSig {
+  private import swift as S
+  private import codeql.swift.controlflow.BasicBlocks as B
+
+  class Location = S::Location;
+
+  class BasicBlock instanceof B::BasicBlock {
+    string toString() { result = super.toString() }
+
+    Callable getEnclosingCallable() { result = super.getScope() }
+
+    Location getLocation() { result = super.getLocation() }
+  }
+
+  BasicBlock getImmediateBasicBlockDominator(BasicBlock bb) {
+    result.(B::BasicBlock).immediatelyDominates(bb)
+  }
+
+  BasicBlock getABasicBlockSuccessor(BasicBlock bb) { result = bb.(B::BasicBlock).getASuccessor() }
+
+  //TODO: support capture of `this` in lambdas
```

Comment:
If I happened to have got the semantics of a `this` reference in a swift lambda wrong in my comments above, then this todo is irrelevant. But if I got it right, then the todo is fine and indeed relevant.

Reply:
Unless of course the `this` variable in Swift is a `VarDecl` in which case you might already be handling this-capture?

Reply:
Oh, I had missed this TODO comment. I believe `this` is a `VarDecl` in Swift, but let me double check that real quick 👀

Reply:
Indeed, a `self` (Swift is one of those cool Smalltalk-like languages 😎) is a `VarDecl`, and a `CapturedDecl` exists for `self` when it's captured in a closure. So this TODO is irrelevant.

Reply:
Do you have a test case for a captured `this`? Something like https://github.com/github/codeql/blob/53d92d58fc27f37888862593dfb6adbd3986aae6/java/ql/test/library-tests/dataflow/capture/B.java#L20-L36 (those missing results in that Java test is what it's about).

Reply:
I've removed the TODO in 78e08cf63ccadde6cdb7f478421460bc816dcf7a. I don't think there's such a test, but I'll make sure to add such a test. There might be something else that's ruining this flow 😂.

Reply:
I've added a version of the above test in 784bb72b3340070a3525a18e7947dfeb3562598b. It turns out we don't properly capture `self`, but this is because of an extractor issue (I've created an internal issue for this now).